### ISSUE TITLE
Want tips for tuning parameters for larges scenes & anti-aliasing

### ISSUE THREAD

**Issue by HrsPythonix** [Totoro97/f2-nerf#34]
Hello, I am wondering if I use f2-NeRF on a large scale dataset, which parameters should be tuned and how to(eg. increase, or reduce).

Also, in my result, I found thin stuff often missing, like image below, is it caused by aliasing?
![image](https://user-images.githubusercontent.com/16586180/236991755-9960c223-64ea-4011-86e6-4e3025951dc2.png)

In your future work part, I see you plan to add anti-aliasing, how do you plan to add it? like ZipNeRF?

**Comment by Totoro97** [comment]
Hi, I think the most critical parameters should be the number of training steps and the hash table size. Two fit larger scenes, I would first try to increase those two parameters. And if larger interaction steps are used, those step-related parameters can also be correspondingly increased. (e.g., subdivision milestones of octree)

Reconstructing thin objects is challenging, aliasing could be one of the reasons though I personally would not think it is the main cause. Decreasing the sampling step size or using L2 loss for the color reconstruction loss may help reconstruct thin objects. (In the default setting of f2-nerf, we use something like smooth-L1 loss)

Regarding anti-aliasing, my current idea for that is very naive, and I haven't thought out specific implementation details. The basic idea is to learn a 4D field (XYZ + volume size) instead of 3D field (XYZ) for volume rendering. But I am not sure if it can work well.

**Comment by HrsPythonix** [comment]
> Hi, I think the most critical parameters should be the number of training steps and the hash table size. Two fit larger scenes, I would first try to increase those two parameters. And if larger iteration steps are used, those step-related parameters can also be correspondingly increased. (e.g., subdivision milestones of octree)
> 
> Reconstructing thin objects is challenging, aliasing could be one of the reasons though I personally would not think it is the main cause. Decreasing the sampling step size or using L2 loss for the color reconstruction loss may help reconstruct thin objects. (In the default setting of f2-nerf, we use something like smooth-L1 loss)
> 
> Regarding anti-aliasing, my current idea for that is very naive, and I haven't thought out specific implementation details. The basic idea is to learn a 4D field (XYZ + volume size) instead of 3D field (XYZ) for volume rendering. But I am not sure if it can work well.

Thanks, I will try it.

**Comment by HrsPythonix** [comment]
> Hi, I think the most critical parameters should be the number of training steps and the hash table size. Two fit larger scenes, I would first try to increase those two parameters. And if larger iteration steps are used, those step-related parameters can also be correspondingly increased. (e.g., subdivision milestones of octree)
> 
> Reconstructing thin objects is challenging, aliasing could be one of the reasons though I personally would not think it is the main cause. Decreasing the sampling step size or using L2 loss for the color reconstruction loss may help reconstruct thin objects. (In the default setting of f2-nerf, we use something like smooth-L1 loss)
> 
> Regarding anti-aliasing, my current idea for that is very naive, and I haven't thought out specific implementation details. The basic idea is to learn a 4D field (XYZ + volume size) instead of 3D field (XYZ) for volume rendering. But I am not sure if it can work well.

How to reduce the sampling size? decrease the sample_l_?

**Comment by Totoro97** [comment]
Hi, decreasing the `sampl_l` can reduce the final sampling size.
Also decreasing the `train.ray_march_init_fineness` (it is a number greater or equal to 1) can reduce the sampling size in the initial stage.
The current implementation performs a coarse-to-fine sampling, i.e., it takes `s * sampl_l` as the sampling step size, where `s` is a number from `ray_marchi_init_finess` to `1` and decreases along the training progression.

**Comment by HrsPythonix** [comment]
> Hi, decreasing the `sampl_l` can reduce the final sampling size. Also decreasing the `train.ray_march_init_fineness` (it is a number greater or equal to 1) can reduce the sampling size in the initial stage. The current implementation performs a coarse-to-fine sampling, i.e., it takes `s * sampl_l` as the sampling step size, where `s` is a number initially as `ray_march_init_finess` as decreases to `1` along the training progression.

Thanks for your reply! I will try to reduce them to see if they take effect.

---
By the way, I implemented L2 loss as you said to replace original L1 Loss, as below:
```
#color_loss = torch::sqrt((pred_colors - gt_colors).square() + 1e-4f).mean();
#to L2 loss
color_loss = (pred_colors - gt_colors).square().mean();
```

The size of my current dataset is 800~1000, I increase the training iters to 100000 by increasing the `train.end_iter`, the result are:
Original L1 loss:
![image](https://github.com/Totoro97/f2-nerf/assets/16586180/6f43e01e-1ab9-4c16-94a3-0c8c3a80bb36)
L2 Loss:
![image](https://github.com/Totoro97/f2-nerf/assets/16586180/84a9b75a-4538-4434-92af-50c46ed12c14)

It seems that the l2 loss can restore more thin objects (not very obvious), and the images rendered with l2 loss looks more 'noisy', not as sharp as l1 loss, though they have similar test PSNR (l1: 22.12 l2: 22.43).

---
In addition, I tried modifying `pts_sampler.sub_div_milestones` in two ways:
```
# original
sub_div_milestones: [ 2000, 4000, 6000, 8000, 10000 ]
# 1
sub_div_milestones: [ 10000, 20000, 30000, 40000, 50000 ]
# 2
sub_div_milestones: [ 2000, 4000, 6000, 8000, 10000, 12000, 14000, 16000, 18000, 20000, 22000, 24000, 26000, 28000, 30000, 32000, 34000, 36000, 38000, 40000, 42000, 44000, 46000, 48000, 50000 ]
```

Their results look similar, I am wondering if I mistakenly use this parameter. 

Can you give some suggestions on these parameters and finetuning? For example, others params such as `pts_sampler.split_dist_thres`, `pts_sampler.scale_by_dis`, are they worth tuning?

**Comment by Totoro97** [comment]
Hi, besides decreasing the sampling steps, I think making the culling condition of the octree node more strict may help preserve the thin structures.

Specifically, you can try to increase the constant `OCC_ALPHA_BASE` in `PtsSampler/PersSampler.cu` from `32` to `64` or more to make the culling logic of the octree nodes not so aggressive.

The culling logic of the octree can be briefly described as follows: During training, we maintain a counter for each node; for each training iteration, if a ray intersects this octree node, we will first check if the maximal alpha value of the sampled point in this node is less than `ABS_ALPHA_THRES`. If it is less than the threshold, we decrease the counter by 1, otherwise, we set the counter to `OCC_ALPHA_BASE`. If the counter is less than zero. We cull out that node. 

This is a parameter I would not usually tune, so I made it hard-coded. But for reconstructing very thin structures, I think this parameter `ABS_ALPHA_THRES` can be increased for a less aggressive culling operation.

**Comment by HrsPythonix** [comment]
> Hi, besides decreasing the sampling steps, I think making the culling condition of the octree node more strict may help preserve the thin structures.
> 
> Specifically, you can try to increase the constant `OCC_ALPHA_BASE` in `PtsSampler/PersSampler.cu` from `32` to `64` or more to make the culling logic of the octree nodes not so aggressive.
> 
> The culling logic of the octree can be briefly described as follows: During training, we maintain a counter for each node; for each training iteration, if a ray intersects this octree node, we will first check if the maximal alpha value of the sampled point in this node is less than `ABS_ALPHA_THRES`. If it is less than the threshold, we decrease the counter by 1, otherwise, we set the counter to `OCC_ALPHA_BASE`. If the counter is less than zero, we cull out that node.
> 
> This is a parameter I would not usually tune, so I made it hard-coded. But for reconstructing very thin structures, I think this parameter `OCC_ALPHA_BASE` can be increased for a less aggressive culling operation.

Thanks for your awesome reply! I will take a try and update results in this issue later.

**Comment by Bin-ze** [comment]
> 嗨，我认为最关键的参数应该是训练步数和哈希表大小。两个适合更大的场景，我会首先尝试增加这两个参数。而如果使用更大的迭代步数，那些与步数相关的参数也可以相应增加。（例如，八叉树的细分里程碑）
> 
> 重建薄对象具有挑战性，混叠可能是原因之一，但我个人认为这不是主要原因。减小采样步长或使用 L2 损失作为颜色重建损失可能有助于重建薄物体。（在 f2-nerf 的默认设置中，我们使用类似 smooth-L1 损失的东西）
> 
> 关于抗锯齿，我目前的想法很幼稚，具体的实现细节也没有想好。基本思想是学习 4D 场（XYZ + 体积大小）而不是 3D 场（XYZ）进行体积渲染。但我不确定它是否能正常工作。

I also encountered the same problem. There are a lot of floating objects in the reconstructed image. How can I modify the configuration file to increase the capacity of the hash table?

**Comment by Bin-ze** [comment]
> > Hi, besides decreasing the sampling steps, I think making the culling condition of the octree node more strict may help preserve the thin structures.
> > Specifically, you can try to increase the constant `OCC_ALPHA_BASE` in `PtsSampler/PersSampler.cu` from `32` to `64` or more to make the culling logic of the octree nodes not so aggressive.
> > The culling logic of the octree can be briefly described as follows: During training, we maintain a counter for each node; for each training iteration, if a ray intersects this octree node, we will first check if the maximal alpha value of the sampled point in this node is less than `ABS_ALPHA_THRES`. If it is less than the threshold, we decrease the counter by 1, otherwise, we set the counter to `OCC_ALPHA_BASE`. If the counter is less than zero, we cull out that node.
> > This is a parameter I would not usually tune, so I made it hard-coded. But for reconstructing very thin structures, I think this parameter `OCC_ALPHA_BASE` can be increased for a less aggressive culling operation.
> 
> Thanks for your awesome reply! I will take a try and update results in this issue later.

Hello, I want to ask you about the progress, I tried to use f2-nerf to reconstruct the indoor scene, but artifacts appeared in many parts, it greatly affects the look and feel, but I don't know how to overcome it, I hope you can give some points suggestion
